m2cgen/interpreters/interpreter.py

```diff
@@ -119,3 +100,33 @@ def interpret_vector_val(self, expr, **kwargs):
         self.with_vectors = True
         nested = [self._do_interpret(expr, **kwargs) for expr in expr.exprs]
         return self._cg.vector_init(nested)
+
+    def _pre_interpret_hook(self, expr, **kwargs):
+        return None, kwargs
+
+    def _do_interpret(self, expr, **kwargs):
+        # Hook which allows to override kwargs and to return custom result.
+        result, kwargs = self._pre_interpret_hook(expr, **kwargs)
+
+        # If result is empty, it means that we still need to process expr.
+        if result is not None:
+            return result
+
+        try:
+            handler = self._select_handler(expr)
+        except NotImplementedError:
+            if isinstance(expr, ast.TransparentExpr):
+                return self._do_interpret(expr.expr,  **kwargs)
+            raise
+
+        if expr.to_cache:
+            if expr not in self._cached_exprs:
+                result = handler(expr, **kwargs)
+                var_name = self._cg.add_var_declaration(expr.output_size)
+                self._cg.add_var_assignment(var_name, result, expr.output_size)
+                self._cached_exprs[expr] = var_name
+                return var_name
+            else:
+                return self._cached_exprs[expr]
+        else:
+            return handler(expr, **kwargs)
```

Comment:
I think this can be simplified a little:

```python
if not expr.to_cache:
    return handler(...)

if expr in self._cached_exprs:
    return self._cached_exprs[expr]

...
```